SW/Boiling Detection/boil_detect.py

```diff
@@ -0,0 +1,240 @@
+        :param thresh: Value signaling if average has fallen beneath certain boiling thresh
+        :return: Boolean signaling if the input list values indicate boiling
+        """
+        average = sum(boiling) / len(boiling)
+        if average < thresh:
+            return True
+        else:
+            return False
+
+    def __burners(self,stove):
+        """
+        Returns a list of the Burners in the Stove object
+        :param stove: Stove object
+        :return: List of the burners
+        """
+        burners = []
+        burners.append(stove.upper_left)
+        burners.append(stove.upper_right)
+        burners.append(stove.lower_left)
+        burners.append(stove.lower_right)
+        return burners
+
+    def __initialize_masks(self,burners):
+        """
+        Refreshes mask if a pot is newly detected on the burners.
+        :param burners: Stove burners
+        """
+        masks = self.masks
+        for mask,burner,quad in zip(masks,burners,self.first_quads):
+            # mask never made before
+            if mask is None and quad is not None and burner.pot_detected:
+                self.masks.append(self.__create_mask(quad))
+            # if pot no longer detected resets the mask
+            elif burner.pot_detected is False:
+                idx = masks.index(mask)
+                self.masks[idx] = None
+
+    def update(self, frame, stove):
+        """
+        Takes an input frame and updates the boiling state of the burners.
+        :param frame: Input frame used to update boiling status
+        :param stove: Stove object holding current stove state
+        """
+        # updates frame_count
+        self.frame_count = self.frame_count + 1
+        burners = self.__burners(stove)
+
+        # creates new masks if a pot was added
+        self.initialize_masks(burners)
+
+        # if program has not executed before
+        if self.first_image is None:
+            self.first_image = frame
+            self.first_quads = self.__split_frame(self.first_image)
+
+        # if the number of frames that have been processed are desired amount, proceed
+        elif self.frame_count == self.count:
+            # if there is no second image recorded yet
+            if self.second_image is None:
+                self.second_image = frame
+                self.second_quads = self.__split_frame(self.second_image)
+            # second image does exist, must shift images
+            else:
+                self.first_image = self.second_image
+                self.second_image = frame
+
+            # splits the frames
+            self.first_quads = self.__split_frame(self.first_image)
+            self.second_quads = self.__split_frame(self.second_image)
+
+            # Detects boiling on active burners
+            i = 0
+            for burner in burners:
+                # if there is a pot on burner
+                if burner.pot_detected:
+                    mask = self.masks[i]
```

Comment:
so this means that the mask is never changed -- is that correct?

maybe it would be better to change the mask if the difference is above some sort of threshold -- does that make sense?

Reply:
right now the mask is changed whenever a new pot is detected on the stove. I'm a little confused on how it would be better to change the mask above a certain threshold?

Reply:
If the pot is placed on the stove and then moved. At the least, you'll probably want a buffer -- like, calculate the threshold after 15 seconds of it being on the stove.

Otherwise, it's going to calculate that threshold the second it comes into the camera view, which will probably be as it's being placed.

Reply:
Believe I fixed this by breaking out of the update function when the calculated SSIM is below MASK_THRESH